modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -512,80 +504,81 @@ private boolean affinityInfoIsUpToDate(int cacheId) {
         return ranges.stream()
             .flatMap(r -> IntStream
                 .rangeClosed(r.portFrom(), r.portTo()).boxed()
-                .map(p -> new InetSocketAddress(r.host(), p))
+                .map(p -> InetSocketAddress.createUnresolved(r.host(), p))
             )
-            .collect(Collectors.toList());
+            .collect(Collectors.toMap(a -> a, a -> 1, Integer::sum));
     }
 
-    /** */
-    private synchronized ClientChannel channel() {
-        if (closed)
-            throw new ClientException("Channel is closed");
-
+    /**
+     * Roll current default channel if specified holder equals to it.
+     */
+    private void rollCurrentChannel(ClientChannelHolder hld) {
+        curChannelsGuard.writeLock().lock();
         try {
-            return channels[curChIdx].getOrCreateChannel();
-        }
-        catch (ClientConnectionException e) {
-            rollCurrentChannel();
+            int idx = curChIdx;
+            List<ClientChannelHolder> holders = channels;
 
-            throw e;
-        }
-    }
+            ClientChannelHolder dfltHld = holders.get(idx);
 
-    /** */
-    private synchronized void rollCurrentChannel() {
-        if (++curChIdx >= channels.length)
-            curChIdx = 0;
+            if (dfltHld == hld) {
+                idx += 1;
+
+                if (idx >= holders.size())
+                    curChIdx = 0;
+                else
+                    curChIdx = idx;
+            }
+        } finally {
+            curChannelsGuard.writeLock().unlock();
+        }
     }
 
     /**
      * On current channel failure.
      */
-    private synchronized void onChannelFailure(ClientChannel ch) {
+    private void onChannelFailure(ClientChannel ch) {
         // There is nothing wrong if curChIdx was concurrently changed, since channel was closed by another thread
         // when current index was changed and no other wrong channel will be closed by current thread because
         // onChannelFailure checks channel binded to the holder before closing it.
-        onChannelFailure(channels[curChIdx], ch);
-
-        chFailLsnrs.forEach(Runnable::run);
+        onChannelFailure(channels.get(curChIdx), ch);
     }
 
     /**
      * On channel of the specified holder failure.
      */
-    private synchronized void onChannelFailure(ClientChannelHolder hld, ClientChannel ch) {
-        if (ch == hld.ch && ch != null) {
+    private void onChannelFailure(ClientChannelHolder hld, ClientChannel ch) {
+        if (ch != null && ch == hld.ch)
             hld.closeChannel();
 
-            if (hld == channels[curChIdx])
-                rollCurrentChannel();
-        }
+        chFailLsnrs.forEach(Runnable::run);
+
+        if (scheduledChannelsReinit.get())
```

Comment:
Perhaps `&& !partitionAwarenessEnabled` should be added here (to avoid double reinit for partition awareness, first time here and second time in async thread). WDYT?

Reply:
Yeah, it is a correct optimization. But I will add this condition inside if block to avoid additional rolling of current channel.

Reply:
fixed